### PR TITLE
feat: catalog filters

### PR DESCRIPTION
**Pull request by magnew**
Closes: #10215 

### Description:

Add filters for search to the data catalog. A couple notes:
- This is done in this review by filtering the post-database-search results. Adding these filters to the database search can probably be done, but requires more
- Filtering is done on the FE, but it could be moved to the BE
- I mixed filtering into the tree build, which makes it harder to separate, but faster. We could also do it as a filter pass on the sorted tree. 
- There are some UX questions. Feedback welcome :D 

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

**Comment by joaoviana** [comment]
To address on this PR or later (up to you!)
1. I would like to be able to revert the filters back to default
2. I would like to know that there are filters applied when the popover is closed. (different color?)

**Comment by rephus** [comment]
Should we filter out `staging` grouped table here ? 

image

**Comment by magnew** [comment]
> To address on this PR or later (up to you!)
> 
> 1. I would like to be able to revert the filters back to default
> 2. I would like to know that there are filters applied when the popover is closed. (different color?)

Cool, Priyanka gave some ideas here. Implementing.

**Comment by magnew** [comment]
> Should we filter out `staging` grouped table here ?
> 
> image

Hmm. Clearly I missed something. Looking.

**Comment by magnew** [comment]
> Should we filter out `staging` grouped table here ?
> 
> image

Could this have been not clicking the apply button? We have now removed it

**Comment by rephus** [comment]
> > Should we filter out `staging` grouped table here ?
> > image
> 
> Could this have been not clicking the apply button? We have now removed it

It's not the case. I can only see a description change. 

[Screencast from 30-05-24 17:00:10.webm](https://github.com/lightdash/lightdash/assets/1983672/f48eeb85-44e1-4c93-a75b-104764bc9dfd)

**Comment by magnew** [comment]
> > Could this have been not clicking the apply button? We have now removed it
> 
> It's not the case. I can only see a description change.
> 
>  Screencast.from.30-05-24.17.00.10.webm

Cool, thanks. I got it. Wasnt hiding tables with only fields and no table returned

**Comment by magnew** [comment]
> I'm not sure why these fields are still appearing here, if the description filter is disabled.
> 
> image

There was confusion about how descriptions should work and Priyanka and I decided to remove it for now.

File: packages/frontend/src/features/catalog/components/CatalogTree.tsx [hunk 1]
```diff
@@ -127,6 +134,9 @@ export const CatalogTree: FC<React.PropsWithChildren<Props>> = ({
         return null;
     }
 
+    const hideGroupedTables =
+        Object.keys(tree).length === 1 && tree['Ungrouped tables'];
```
Comment (joaoviana) [inline packages/frontend/src/features/catalog/components/CatalogTree.tsx:138]: sorry I don't get this. Can you explain?

File: packages/frontend/src/features/catalog/components/CatalogPanel.tsx [hunk 2]
```diff
@@ -96,6 +100,20 @@ function getSelectionList(tree: CatalogTreeType): CatalogSelection[] {
     return selectionList;
 }
 
+enum FilterType {
+    Dimensions = 'dimensions',
+    Metrics = 'metrics',
+    HideBaseTables = 'hideBaseTables',
```
Comment (joaoviana) [inline packages/frontend/src/features/catalog/components/CatalogPanel.tsx:106]: maybe we can remove this until implemented?

File: packages/frontend/src/features/catalog/components/CatalogPanel.tsx [hunk 3]
```diff
@@ -385,33 +472,176 @@ export const CatalogPanel: FC = () => {
...
+                                        );
+                                    }}
+                                    label={
+                                        <Text fz="xs" fw={500} c="gray.7">
+                                            Hide grouped tables
+                                        </Text>
+                                    }
+                                />
+                            </Stack>
+                            <Divider c="gray.1" />
```
Comment (joaoviana) [inline packages/frontend/src/features/catalog/components/CatalogPanel.tsx:627]: I know it's in the designs but I would remove this - feel free to ignore - no change required tbh!

**Review by joaoviana — APPROVED** [review]
Working great for me! Some considerations but non-blocking

File: packages/frontend/src/features/catalog/components/CatalogPanel.tsx (hunk 2 above)
Comment (magnew) [reply packages/frontend/src/features/catalog/components/CatalogPanel.tsx:106]: Sure!

File: packages/frontend/src/features/catalog/components/CatalogPanel.tsx (hunk 3 above)
Comment (magnew) [reply packages/frontend/src/features/catalog/components/CatalogPanel.tsx:627]: Good catch

File: packages/frontend/src/features/catalog/components/CatalogTree.tsx (hunk 1 above)
Comment (magnew) [reply packages/frontend/src/features/catalog/components/CatalogTree.tsx:138]: If there is only one group and it is the ungrouped tables group, we don't show the group name (which would be 'Ungrouped tables'). Hiding this wasn't in the designs but it felt weird to have one unnamed node there.

File: packages/frontend/src/features/catalog/components/CatalogPanel.tsx [hunk 4]
```diff
@@ -385,33 +472,176 @@ export const CatalogPanel: FC = () => {
...
+                                                },
+                                            }}
+                                        >
+                                            Hide grouped tables
+                                        </Badge>
+                                    )}
+                                </Group>
+                            </Button>
+                        </Popover.Target>
+                        <Button
```
[the remark this answers is not in the document]
Comment (magnew) [reply packages/frontend/src/features/catalog/components/CatalogPanel.tsx:540]: Yea, good call

**Pull request merged**